### PR TITLE
support ipxe scripts over data uris in ipxe script urls

### PR DESCRIPTION
**Pull request by displague**
## Description

from https://github.com/tinkerbell/boots/pull/192#discussion_r687810189

## Why is this needed

Fixes: #110

## How Has This Been Tested?

Tests are failing, I haven't looked into how to fix this yet. I think the existing mocks may not be sufficient.

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: installers/custom_ipxe/main_test.go [hunk 1]
```diff
@@ -130,6 +130,25 @@ func TestIpxeScript(t *testing.T) {
...
+			params
+			param body Device connected to DHCP system
+			param type provisioning.104.01
+			imgfetch ${tinkerbell}/phone-home##params
+			imgfree
+
+			set packet_facility test.facility
+			set packet_plan test.slug
+			
+			echo my data uri script
```
Comment (displague) [inline installers/custom_ipxe/main_test.go:149]: The test fails, with this line being a chain-url line, referencing the unexpanded data uri.

File: installers/custom_ipxe/main_test.go [hunk 2]
```diff
@@ -130,6 +130,25 @@ func TestIpxeScript(t *testing.T) {
 			echo userdata script
 			`,
 		},
+		{
+			"installer: ipxe data url",
+			"custom_ipxe",
+			&packet.InstallerData{Chain: "data:text/plain;charset=utf-8;base64,ZWNobyBteSBkYXRhIHVyaSBzY3JpcHQ="},
```
Comment (displague) [inline installers/custom_ipxe/main_test.go:136]: this data uri is: `echo my data uri script`

**Comment by displague** [comment]
@jeremytanner I would like to see this work continued for #110. I am stalled on it.

@jacobweinstock do you see what I'd like to have here (based on the test)? Is this the right place for this change and tests?

**Comment by jacobweinstock** [comment]
Hey @displague, apologies for the lack of response here. I will put this on my priority list for this week/next and update here when I have something concrete to provide. Again, sorry for the delay.

File: installers/custom_ipxe/main_test.go (hunk 2 above)
Comment (tstromberg) [reply installers/custom_ipxe/main_test.go:136]: Since this comment reduced the mystery for me, can you add this as a comment in the code base?

**Review by tstromberg — DISMISSED** [review]
Minor comments.

File: installers/custom_ipxe/main_test.go (hunk 1 above)
Comment (tstromberg) [reply installers/custom_ipxe/main_test.go:149]: Can you paste the test failure here?

**Comment by tstromberg** [comment]
PR overall looks good - is it just stuck on there being a failing test?

File: installers/custom_ipxe/main_test.go (hunk 1 above)
Comment (displague) [reply installers/custom_ipxe/main_test.go:149]: I'm rerunning the test to capture the log output.

Comment (displague) [reply installers/custom_ipxe/main_test.go:151]:
```
--- FAIL: TestIpxeScript (0.00s)
    --- FAIL: TestIpxeScript/installer:_ipxe_data_url (0.00s)
        main_test.go:172:
            	Error Trace:	main_test.go:172
            	Error:      	Not equal:
            	            	expected: "#!ipxe\n\n\nparams\nparam body Device connected to DHCP system\nparam type provisioning.104.01\nimgfetch ${tinkerbell}/phone-home##params\nimgfree\n\nset packet_facility test.facility\nset packet_plan test.slug\n\necho my data uri script\n"
            	            	actual  : "#!ipxe\n\n\nparams\nparam body Device connected to DHCP system\nparam type provisioning.104.01\nimgfetch ${tinkerbell}/phone-home##params\nimgfree\n\nset packet_facility test.facility\nset packet_plan test.slug\nchain --autofree data:text/plain;charset=utf-8;base64,ZWNobyBteSBkYXRhIHVyaSBzY3JpcHQ=\n"

            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -11,4 +11,3 @@
            	            	 set packet_plan test.slug
            	            	+chain --autofree data:text/plain;charset=utf-8;base64,ZWNobyBteSBkYXRhIHVyaSBzY3JpcHQ=

            	            	-echo my data uri script
            	            	-
            	Test:       	TestIpxeScript/installer:_ipxe_data_url
{"level":"error","ts":1637329705.027709,"caller":"custom_ipxe/main.go:54","msg":"validating ipxe config","service":"github.com/tinkerbell/boots","error":"ipxe config URL or Script must be defined"}
FAIL
```

**Comment by crayzeigh** [comment]
Based on comment history it looks like this one is still awaiting @jacobweinstock to prioritize and review. Assigning failing test label based on description

**Comment by jacobweinstock** [comment]
@displague, looks like this is still in draft and has some merge conflicts. will wait for updates before reviewing?

**Pull request closed without merging**